pkg/kv/kvserver/logstore/sideload.go

```diff
@@ -140,8 +143,12 @@ func MaybeSideloadEntries(
 		sideloadedEntriesSize += int64(len(dataToSideload))
 	}
 
-	if output == nil {
-		// We never saw a sideloaded command.
+	if output != nil { // there is at least one sideloaded command
+		// Sync the sideloaded storage directory so that the commands are durable.
+		if err := sideloaded.Sync(); err != nil {
```

Comment:
This sync could be moved up the stack, to `LogStore.storeEntriesAndCommitBatch` or higher if we want a delayed/"async" `Sync`. The sideloaded storage syncs can't be batched across ranges/replicas though (because each one is on a separate directory), so there is not much sense in moving it up right now.

Reply:
Each `Put()` syncs the added file, but not the containing directory.

`Sync()` syncs the directory containing these files (so the references to the newly added files are persisted), and is called once per `storeEntriesAndCommitBatch` (which is called once per raft ready run). I.e. it "batches" things to some extent, if there are multiple `Put()` entries in a batch.

Other directories (parents of the per-range directories; i.e. the entire path like `auxiliary/sideloading/r2XXX`) are synced once in `mkdirAllAndSyncParents` when they are created. After that, they don't need to be synced again.

Reply:
I don't understand. Consider this sequence, using `auxiliary/sideloaded` as the base directory:

```
Put("r1XXX/r1234")
Sync()
```

`Sync()` will only sync `auxiliary/sideloaded`, so it won't guarantee durability of `r1XXX/r1234`'s entry in `r1XXX`, only `r1XXX`'s entry in the base dir.

If `Put()` is supposed to guarantee durability of `r1XXX/r1234`'s directory entry, then it should presumably also guarantee durability of `somefile` in the base dir, in which case `Sync()` is redundant.

What am I missing?

Reply:
Nevermind, we instantiate the sideloaded storage per replica, so it's always a flat directory.

Reply:
The base directory is `auxiliary`, and the full path to the sideloaded storage for range 1234 is `auxiliary/sideloaded/r1XXX/r1234`.

When a sideloaded storage for range 1234 is created, it needs to create and sync the entire `auxiliary/sideloaded/r1XXX/r1234` path. But we don't do it instantly on creation, instead it's done lazily when `Put` writes a first entry file. If `Put` discovers that the directories are missing, it creates and syncs them. But we can pretty much assume that it happens only once.

The normal operation of `Put` only adds files to an already existing `auxiliary/sideloaded/r1XXX/r1234` directory. It does so by calling [WriteFileSyncing](https://github.com/cockroachdb/cockroach/blob/dcedaf1cc617796bce6f8313f74631e1c0dbcad8/pkg/kv/kvserver/logstore/sideload_disk.go#L100) which fsyncs the file's content.

After we've run a bunch of `Put`s, everything is synced except the `.../r1234` folder (aka the [dir](https://github.com/cockroachdb/cockroach/blob/dcedaf1cc617796bce6f8313f74631e1c0dbcad8/pkg/kv/kvserver/logstore/sideload_disk.go#L42C1-L42C1) field), because it now contains some new files. That's why we need the additional `Sync()` method, so sync this `ss.dir` folder.

Reply:
I split out a couple more commits to demonstrate the effect of multiple syncs in this PR. See the evolution of `TestSideloadStorageSync` across the commits.